### PR TITLE
Back-end collapse API - Convert generic JSON objects to Java POJOs for strong typing

### PR DESCRIPTION
**Pull request by joeferner**
- [x] joeferner
- [x] sfeng88
- [x] mwizeman joeybrk372 jharwig

This commit gives API users a type safe interface to collapse entities.

CHANGELOG
Changed: Back-end collapse API - Convert generic JSON objects to Java POJOs for strong typing

### CODE REVIEW

**Comment by joeferner** [comment]
On my project I am generating workspaces and I need a way to collapse vertices using the new API. Typing in weakly types JSON seemed very fragile.

**Comment by jharwig** [comment]
@joeferner would this prevent plugins from adding to extendedData?

**Comment by joeferner** [comment]
@jharwig It works with the graph styling plugin (I assume it should work with others as well). Vertices, edges, and compound nodes are a little different in the extended data structure in that they aren't stored anywhere but are computed when asked for so this change doesn't effect the other parts of the extended data because of this.

Existing plugin extended data is retrieved in the getProductExtendedDataJson method https://github.com/visallo/visallo/pull/1309/files#diff-a0f439d5c2ad6a598890917b6f2915daR893 which still loads the property as JSON and appends to the JSON.

The saving is handled here https://github.com/visallo/visallo/blob/bd264f6b90f702c78da05ffc3046233af65962f3/core/plugins/model-vertexium/src/main/java/org/visallo/vertexium/model/workspace/VertexiumWorkspaceRepository.java#L1073 which isn't effected by these changes.

File: core/core/src/main/java/org/visallo/core/model/workspace/product/WorkProductVertex.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+import org.visallo.web.clientapi.model.ClientApiObject;
+import org.visallo.web.clientapi.model.GraphPosition;
+
+import java.util.List;
+
+public class WorkProductVertex implements ClientApiObject {
+    private String id;
+    private boolean visible;
+    private String parent;
```
Comment (jharwig) [inline core/core/src/main/java/org/visallo/core/model/workspace/product/WorkProductVertex.java]: These are specific to graph. How do other products changes these?

File: web/plugins/graph-product/src/main/java/org/visallo/web/product/graph/GraphWorkProductService.java [hunk 2]
```diff
@@ -528,52 +522,57 @@ private GraphPosition calculatePositionFromParents(
...
+            if (children != null) {
+                GraphProductOntology.NODE_CHILDREN.updateProperty(elemCtx, children, visibility);
+            }
         }
     }
 
-    protected void setEdgeJson(Edge propertyVertexEdge, JSONObject vertex) {
-        JSONObject position = ENTITY_POSITION.getPropertyValue(propertyVertexEdge);
+    @Override
+    protected void populateVertexWithWorkspaceEdge(Edge propertyVertexEdge, WorkProductVertex vertex) {
```
Comment (jharwig) [inline web/plugins/graph-product/src/main/java/org/visallo/web/product/graph/GraphWorkProductService.java]: Maybe we can change the signature to return a `WorkProductVertex` subclass here like `GraphWorkProductVertex`?

Comment (joeferner) [reply web/plugins/graph-product/src/main/java/org/visallo/web/product/graph/GraphWorkProductService.java]: I moved parent, pos, and children to a subclass.  Leaving id, visible, title, type, unauthorized in the base class.

**Comment by jharwig** [comment]
Go ahead and merge after travis succeeds

**Pull request merged**